server/src/main/java/org/eclipse/lsp/cobol/core/engine/SyntaxTreeEngine.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright (c) 2020 Broadcom.
+ * The term "Broadcom" refers to Broadcom Inc. and/or its subsidiaries.
+ *
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *    Broadcom, Inc. - initial API and implementation
+ *
+ */
+package org.eclipse.lsp.cobol.core.engine;
+
+import org.antlr.v4.runtime.Token;
+import org.eclipse.lsp.cobol.core.messages.MessageService;
+import org.eclipse.lsp.cobol.core.model.Locality;
+import org.eclipse.lsp.cobol.core.model.SyntaxError;
+import org.eclipse.lsp.cobol.core.model.tree.*;
+import org.eclipse.lsp.cobol.core.model.variables.Variable;
+import org.eclipse.lsp.cobol.core.visitor.VariableDefinitionDelegate;
+import org.eclipse.lsp.cobol.core.visitor.VariableUsageDelegate;
+
+import java.util.*;
+
+/**
+ * This class is responsible for processing the syntax tree.
+ */
+public class SyntaxTreeEngine {
+  private final Node rootNode;
+  private final Map<Token, Locality> positionMapping;
+  private final MessageService messageService;
+  private final List<SyntaxError> errors = new ArrayList<>();
+
+  public SyntaxTreeEngine(Node rootNode, Map<Token, Locality> positionMapping, MessageService messageService) {
+    this.rootNode = rootNode;
+    this.positionMapping = positionMapping;
+    this.messageService = messageService;
+  }
+
+  /**
+   * Runs the process under the syntax tree.
+   */
+  public void processTree() {
+    rootNode.getDepthFirstStream()
+        .filter(it -> it.getNodeType() == NodeType.PROGRAM)
+        .forEach(it -> processProgram((ProgramNode) it));
+  }
+
+  private void processProgram(ProgramNode programNode) {
```

Comment:
How did you integrate the section swiching logic here?

Reply:
I go through all nodes and if I meet a section, I call `variablesDelegate.notifySectionChanged()`.
The behavior is exactly the save we have in CobolVisitor right now.